research/delf/delf/python/training/global_features/train_utils.py

```diff
@@ -0,0 +1,368 @@
+  losses = global_features_utils.AverageMeter()
+
+  # Retrieve all trainable variables we defined in the graph.
+  tvs = model.trainable_variables
+  accum_grads = [tf.Variable(tf.zeros_like(tv.read_value()), trainable=False)
+                 for tv in tvs]
+
+  end = time.time()
+  batch_num = 0
+  print_frequency = 10
+  all_batch_num = query_size // batch_size
+  state = 'Train' if train else 'Val'
+  global_features_utils.debug_and_log('>> {} step:'.format(state))
+
+  # For every batch in the dataset; Stops when all batches in the dataset have
+  # been processed.
+  while True:
+    data_time.update(time.time() - end)
+
+    if train:
+      try:
+        # Train on one batch.
+        # We load batches into memory consequently.
+        for _ in range(batch_size):
+          # Because the images are not necessarily of the same size, we can't
+          # set the batch size with .batch().
+          batch = loader.get_next()
+          input_tuple = batch[0:-1]
+          target_tuple = batch[-1]
+
+          loss_value, grads = grad(criterion, model, input_tuple,
+                                   target_tuple, neg_num)
+          losses.update(loss_value)
+          # Adds to each element from the list you initialized earlier
+          # with zeros its gradient (works because accum_vars and gvs
+          # are in the same order).
+          accum_grads = [accum_grads[i].assign_add(gv) for i, gv in
+                         enumerate(grads)]
+
+        if (batch_num + 1) % update_every == 0 or (
+                batch_num + 1) == all_batch_num:
+          # Do one step for multiple batches. Accumulated gradients are
+          # used.
+          optimizer.apply_gradients(
+                  zip(accum_grads, model.trainable_variables))
+          accum_grads = [
+            tf.Variable(tf.zeros_like(tv.read_value()), trainable=False)
+            for tv in tvs]
+      except Exception as ex:
+        global_features_utils.debug_and_log(ex)
+        break
+
+    else:
+      # Validate one batch.
+      # We load full batch into memory.
+      input = []
+      target = []
+      try:
+        for _ in range(batch_size):
+          # Because the images are not necessarily of the same size, we can't
+          # set the batch size with .batch().
+          batch = loader.get_next()
+          input.append(batch[0:-1])
+          target.append(batch[-1])
+      except Exception as ex:
+        global_features_utils.debug_and_log(ex)
+        break
+
+      output = tf.zeros(shape=(0, model.meta['outputdim']), dtype=tf.float32)
+      for input_tuple in input:
+        for img in input_tuple:
+          # Compute the global descriptor vector.
+          model_out = model(tf.expand_dims(img, axis=0), training=False)
+          output = tf.concat([output, model_out], 0)
+
+      # No need to reduce memory consumption (no backward pass):
+      # Compute loss for the full batch.
+      tmp_target = tf.concat(target, axis=0)
```

Comment:
trying to understand, is `tmp_target` needed at all? Why can't we just use `target` directly? Sorry if I missed sthing.